File: electron/menu.ts
// 原生菜单 + 加速键。菜单项只负责把对应 MenuAction 经 webContents.send
// 发给渲染端,具体行为(新建/打开/保存…)由壳层接 onMenuAction 决定,
// 主进程不在这里读写文件——保持"菜单=纯触发器"。

import { app, Menu, type BrowserWindow, type MenuItemConstructorOptions } from "electron";
import type { MenuAction } from "../src/shell/ipc";

const isMac = process.platform === "darwin";

// buildMenu 收一个取窗口的 getter(窗口可能在重建中为 null),
// 每次点击时现取 webContents,避免持有过期引用。
export function buildMenu(getWin: () => BrowserWindow | null): Menu {
  const send = (action: MenuAction) => {
    getWin()?.webContents.send("mkn:menu-action", action);
  };

  const fileMenu: MenuItemConstructorOptions = {
    label: "文件",
    submenu: [
      { label: "新建", accelerator: "CmdOrCtrl+N", click: () => send("new") },
      { label: "打开…", accelerator: "CmdOrCtrl+O", click: () => send("open") },
      {
        label: "打开文件夹…",
        accelerator: "Shift+CmdOrCtrl+O",
        click: () => send("openFolder"),
      },
      { type: "separator" },
      { label: "保存", accelerator: "CmdOrCtrl+S", click: () => send("save") },
      {
        label: "另存为…",
        accelerator: "Shift+CmdOrCtrl+S",
        click: () => send("saveAs"),
      },
      { type: "separator" },
      {
        // 导出动作只下发 MenuAction;HTML 生成 / pandoc 可用性判断都在渲染端壳层。
        // "导出 Word" 常显:渲染端先 hasPandoc() 再决定是否提示安装,菜单不预先禁用。
        label: "导出",
        submenu: [
          { label: "导出 HTML…", click: () => send("exportHtml") },
          { label: "导出 PDF…", click: () => send("exportPdf") },
          { label: "导出 Word(pandoc)…", click: () => send("exportDocx") },
        ],
      },
      { type: "separator" },
      // 查找走渲染端的 CodeMirror 搜索面板;⌘F 统一在这里截获再下发。
      { label: "查找", accelerator: "CmdOrCtrl+F", click: () => send("find") },
      { type: "separator" },
      isMac ? { role: "close", label: "关闭窗口" } : { role: "quit", label: "退出" },
    ],
  };

  const editMenu: MenuItemConstructorOptions = {
    label: "编辑",
    submenu: [
      { role: "undo", label: "撤销" },
      { role: "redo", label: "重做" },
      { type: "separator" },
      { role: "cut", label: "剪切" },
      { role: "copy", label: "复制" },
      { role: "paste", label: "粘贴" },
      { role: "selectAll", label: "全选" },
    ],
  };

  const viewMenu: MenuItemConstructorOptions = {
    label: "视图",
    submenu: [
      {
        label: "切换侧栏",
        accelerator: "CmdOrCtrl+\\",
        click: () => send("toggleSidebar"),
      },
      { type: "separator" },
      { role: "resetZoom", label: "实际大小" },
      { role: "zoomIn", label: "放大" },
      { role: "zoomOut", label: "缩小" },
      { type: "separator" },
      { role: "togglefullscreen", label: "全屏" },
      // 开发期排障用;打包后保留无妨(不主动弹出)。
      { role: "toggleDevTools", label: "开发者工具" },
    ],
  };

  const windowMenu: MenuItemConstructorOptions = {
    label: "窗口",
    role: "windowMenu",
  };

  const template: MenuItemConstructorOptions[] = [];

  // macOS 首菜单为应用名,含"关于/退出"等标准项。
  if (isMac) {
    template.push({
      label: app.name,
      submenu: [
        { role: "about", label: `关于 ${app.name}` },
        { type: "separator" },
        { role: "hide", label: `隐藏 ${app.name}` },
        { role: "hideOthers", label: "隐藏其他" },
        { role: "unhide", label: "全部显示" },
        { type: "separator" },
        { role: "quit", label: `退出 ${app.name}` },
      ],
    });
  }

  template.push(fileMenu, editMenu, viewMenu, windowMenu);

  return Menu.buildFromTemplate(template);
}

File: src-tauri/src/menu.rs
//! 原生菜单 + 加速键。对齐 electron/menu.ts:菜单项只负责把对应 MenuAction
//! emit 给渲染端,具体行为由壳层接 onMenuAction 决定,后端不在这里读写文件。
//!
//! 与 Electron 差异:zoom / devTools 是 Electron role,Tauri 无对应预定义项,
//! 且不属于 MknApi 契约(纯装饰),移植时省略;其余结构/加速键逐一对齐。

use tauri::menu::{Menu, MenuBuilder, MenuItemBuilder, PredefinedMenuItem, SubmenuBuilder};
use tauri::{AppHandle, Emitter, Wry};

/// 自定义菜单项 id 即 MknApi 的 MenuAction 字符串,on_menu_event 直接转发。
const ACTIONS: &[&str] = &[
    "new",
    "open",
    "openFolder",
    "save",
    "saveAs",
    "find",
    "toggleSidebar",
    "exportHtml",
    "exportPdf",
    "exportDocx",
];

pub fn build_menu(app: &AppHandle) -> tauri::Result<Menu<Wry>> {
    let is_mac = cfg!(target_os = "macos");

    let mi = |id: &str, label: &str, accel: Option<&str>| {
        let mut b = MenuItemBuilder::new(label).id(id);
        if let Some(a) = accel {
            b = b.accelerator(a);
        }
        b.build(app)
    };

    // ── 文件 ──
    let export_sub = SubmenuBuilder::new(app, "导出")
        .item(&mi("exportHtml", "导出 HTML…", None)?)
        .item(&mi("exportPdf", "导出 PDF…", None)?)
        .item(&mi("exportDocx", "导出 Word(pandoc)…", None)?)
        .build()?;

    let mut file = SubmenuBuilder::new(app, "文件")
        .item(&mi("new", "新建", Some("CmdOrCtrl+N"))?)
        .item(&mi("open", "打开…", Some("CmdOrCtrl+O"))?)
        .item(&mi("openFolder", "打开文件夹…", Some("Shift+CmdOrCtrl+O"))?)
        .separator()
        .item(&mi("save", "保存", Some("CmdOrCtrl+S"))?)
        .item(&mi("saveAs", "另存为…", Some("Shift+CmdOrCtrl+S"))?)
        .separator()
        .item(&export_sub)
        .separator()
        .item(&mi("find", "查找", Some("CmdOrCtrl+F"))?)
        .separator();
    file = if is_mac {
        file.item(&PredefinedMenuItem::close_window(app, Some("关闭窗口"))?)
    } else {
        file.item(&PredefinedMenuItem::quit(app, Some("退出"))?)
    };
    let file = file.build()?;

    // ── 编辑 ──
    let edit = SubmenuBuilder::new(app, "编辑")
        .item(&PredefinedMenuItem::undo(app, Some("撤销"))?)
        .item(&PredefinedMenuItem::redo(app, Some("重做"))?)
        .separator()
        .item(&PredefinedMenuItem::cut(app, Some("剪切"))?)
        .item(&PredefinedMenuItem::copy(app, Some("复制"))?)
        .item(&PredefinedMenuItem::paste(app, Some("粘贴"))?)
        .item(&PredefinedMenuItem::select_all(app, Some("全选"))?)
        .build()?;

    // ── 视图 ──
    let view = SubmenuBuilder::new(app, "视图")
        .item(&mi("toggleSidebar", "切换侧栏", Some("CmdOrCtrl+\\"))?)
        .separator()
        .item(&PredefinedMenuItem::fullscreen(app, Some("全屏"))?)
        .build()?;

    // ── 窗口 ──
    let window = SubmenuBuilder::new(app, "窗口")
        .item(&PredefinedMenuItem::minimize(app, Some("最小化"))?)
        .item(&PredefinedMenuItem::maximize(app, Some("缩放"))?)
        .build()?;

    let mut menu = MenuBuilder::new(app);

    // macOS 首菜单为应用名,含 关于/隐藏/退出 标准项。
    if is_mac {
        let app_name = "隐墨";
        let app_menu = SubmenuBuilder::new(app, app_name)
            .item(&PredefinedMenuItem::about(app, Some(&format!("关于 {app_name}")), None)?)
            .separator()
            .item(&PredefinedMenuItem::hide(app, Some(&format!("隐藏 {app_name}")))?)
            .item(&PredefinedMenuItem::hide_others(app, Some("隐藏其他"))?)
            .item(&PredefinedMenuItem::show_all(app, Some("全部显示"))?)
            .separator()
            .item(&PredefinedMenuItem::quit(app, Some(&format!("退出 {app_name}")))?)
            .build()?;
        menu = menu.item(&app_menu);
    }

    menu.item(&file)
        .item(&edit)
        .item(&view)
        .item(&window)
        .build()
}

/// 菜单点击 → emit mkn:menu-action(对齐 menu.ts 的 webContents.send)。
pub fn handle_menu_event(app: &AppHandle, id: &str) {
    if ACTIONS.contains(&id) {
        let _ = app.emit("mkn:menu-action", id);
    }
}

